src/driver/hermes_cdev.c

```diff
@@ -97,43 +97,69 @@ static int hermes_close(struct inode *inode, struct file *filp)
 	return 0;
 }
 
-static ssize_t hermes_read(struct file *filp, char __user *buff, size_t count,
-		loff_t *f_pos)
+static ssize_t hermes_read_iter(struct kiocb *iocb, struct iov_iter *to)
 {
-	struct hermes_env *env = filp->private_data;
+	struct hermes_env *env = iocb->ki_filp->private_data;
 	struct hermes_pci_dev *hpdev = env->hermes->hpdev;
 	struct hermes_cfg *cfg = &hpdev->hdev->cfg;
 	struct xdma_channel *chnl;
-	long bytes;
-	loff_t pos, offset = *f_pos;
+	long bytes, res = 0;
+	loff_t pos, offset = iocb->ki_pos;
+	size_t count;
+	int i ;
 
 	if (env->data_slot < 0)
 		return -ENODATA;
 
+	if (iocb->ki_flags & (IOCB_HIPRI | IOCB_NOWAIT))
+		return -EOPNOTSUPP;
+
+	if (offset == -1)
+		return -EOPNOTSUPP;
+
 	chnl = xdma_get_c2h(hpdev);
 
-	if (count + offset > cfg->ehdssze) {
-		count = cfg->ehdssze - offset;
-		if ((int) count < 0)
-			count = 0;
-	}
+	for (i = 0; i < to->nr_segs; i++) {
```

Comment:
Should really be using `iov_iter_get_pages` in a loop, not creating your own loop... I'd probably refactor the xdma code to use it somehow.

Otherwise your making an invalid assumption that the iov contains iovecs and not any of the other types of data it supports.